roachtest: add INSPECT DATABASE post-validation check

spilchen opened the pull request:
Add INSPECT DATABASE as a post-validation step in selected roachtests to verify consistency.

Initially enabled only for alterpk* and schemachange* tests, owned by SQL Foundations. Tests that already run INSPECT internally explicitly skip this step. A regex-based blacklist disables it for most existing tests to allow a gradual rollout and avoid destabilizing test runs.

Controlled via a new PostValidation flag. Future tests are expected to enable this by default.

Fixes #154049
Epic: CRDB-30356
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/155438)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
this approach looks good! my comments are pretty minor

For the final sign-off, I'll defer to TestEng to approve the adding of new validation in the roachtest framework.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-:-ObiPwxC1JgbFFlVPDGk:bf8sbhn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @herkolategan)

___
*[`pkg/cmd/roachtest/roachtestutil/validation_check.go` line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiNh3a4kpIrSpQrw0R:-ObiNh3a4kpIrSpQrw0S:b9zgj3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/roachtestutil/validation_check.go#L190)):*
> ```Go
> 	rows, err := db.QueryContext(ctx, `
> 		SELECT datname FROM pg_database
> 		WHERE datname NOT IN ('system', 'postgres', 'defaultdb')
> ```

perhaps we should only skip `system`. i don't believe it's that uncommon for roachtests to do their work in `defaultdb`, and if they do, we'd want to check that database.

`postgres` and `defaultdb` are "normal" user-facing databases; the only special thing about them is that they exist upon cluster initialization.
___
*[`pkg/cmd/roachtest/roachtestutil/validation_check.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiNOu51EDKrWTaSCL8:-ObiNOu51EDKrWTaSCL9:b-ctr5o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/roachtestutil/validation_check.go#L253)):*
> ```Go
> 			l.Printf("Launching INSPECT DATABASE %s", dbName)
> 			inspectQuery := fmt.Sprintf(
> 				"SET enable_inspect_command = true; SET statement_timeout = '%s'; COMMIT; INSPECT DATABASE %s",
> ```

why do we need the `COMMIT` statement?

also, i think we should quote the database name in case it's mixed-case or has special characters. we can use: `lexbase.EscapeSQLIdent(dbName)`
___
*[`pkg/cmd/roachtest/roachtestutil/validation_check.go` line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiPPmF8d7wQb0LmPV0:-ObiPPmF8d7wQb0LmPV1:bakgpsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/roachtestutil/validation_check.go#L263)):*
> ```Go
> 				// Feature not supported errors indicate the cluster version doesn't support
> 				// INSPECT yet (requires v25.4). Return this error to the caller.
> 				if isFeatureNotSupportedError(err) {
> ```

technically, we don't need this check for `isFeatureNotSupportedError` here. default case of wrapping the error and then checking `isFeatureNotSupportedError` in the caller will also achieve the desired outcome of ignoring these errors.
___
*[`pkg/cmd/roachtest/test_runner.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiOEcP06Uhy10V49F2:-ObiOEcP06Uhy10V49F3:bez6o94) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/test_runner.go#L143)):*
> ```Go
> const VmLabelTestOwner string = "test_owner"
> 
> // inspectBlacklistRegex is a compiled regex of test name patterns that should
> ```

super nit: we generally use the term "blocklist" rather than "blacklist" throughout the project.
___
*[`pkg/cmd/roachtest/test_runner.go` line 1784 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiOuf93uo9YUwSWrWl:-ObiOuf93uo9YUwSWrWm:b-4jy7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/test_runner.go#L1784)):*
> ```Go
> 
> 				db := c.Conn(ctx, t.L(), validationNode,
> 					option.VirtualClusterName(install.SystemInterfaceName))
> ```

is there something we should do to also INSPECT non-system tenants? (test-eng may have thoughts on how to do this or if it's possible.)


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155438)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @rafiss)

___
*[`pkg/cmd/roachtest/test_runner.go` line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiOEcP06Uhy10V49F2:-ObmI13R7LnDRDrbPaHx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/test_runner.go#L143)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: we generally use the term "blocklist" rather than "blacklist" throughout the project.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/test_runner.go` line 1784 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiOuf93uo9YUwSWrWl:-ObmJn5U90LqmLsdLR5u:b-5g6wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/test_runner.go#L1784)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is there something we should do to also INSPECT non-system tenants? (test-eng may have thoughts on how to do this or if it's possible.)
</blockquote></details>

I dropped this VirtualClusterName option as per Darryl's comment.
___
*[`pkg/cmd/roachtest/roachtestutil/validation_check.go` line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiNh3a4kpIrSpQrw0R:-ObmIAm5DqL0JOHdR9fz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/roachtestutil/validation_check.go#L190)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

perhaps we should only skip `system`. i don't believe it's that uncommon for roachtests to do their work in `defaultdb`, and if they do, we'd want to check that database.

`postgres` and `defaultdb` are "normal" user-facing databases; the only special thing about them is that they exist upon cluster initialization.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/roachtestutil/validation_check.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiNOu51EDKrWTaSCL8:-ObmTf-a01K46XM6QTRp:b-qa6qcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/roachtestutil/validation_check.go#L253)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why do we need the `COMMIT` statement?

also, i think we should quote the database name in case it's mixed-case or has special characters. we can use: `lexbase.EscapeSQLIdent(dbName)`
</blockquote></details>

You can't run INSPECT within a transaction, and all of the SQL statements are sent in one shot. I do admit it looks odd. I changed it so that we call `db.ExecContext` separately. 
___
*[`pkg/cmd/roachtest/roachtestutil/validation_check.go` line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObiPPmF8d7wQb0LmPV0:-ObmJ4CTCCQp8i7Ex-6P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ab9bb296537f64e95582adafeefe982c511e46/pkg/cmd/roachtest/roachtestutil/validation_check.go#L263)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

technically, we don't need this check for `isFeatureNotSupportedError` here. default case of wrapping the error and then checking `isFeatureNotSupportedError` in the caller will also achieve the desired outcome of ignoring these errors.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DarrylWong commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -139,6 +140,118 @@ const VmLabelTestRunID string = "test_run_id"
+// inspectBlocklistRegex is a compiled regex of test name patterns that should
+// skip INSPECT validation. Tests matching any of these patterns will not run
+// INSPECT.
+var inspectBlocklistRegex = regexp.MustCompile(
```
A blocklist seems okay to me if the plan is for this to be a relatively quick rollout. Otherwise would a allowlist make more sense? That way we don't run into a state where we have to remember to add new tests to this list.

spilchen replied:
New tests should never have to update this list. There is a skip flag if they don't want the INSPECT validation. This list just exists to avoid a "thundering herd" of all of the roachtest doing INSPECT validation in the same batch. It allows us to roll out the inspect validation in a more staggered approach. I hope to come back to this and start removing test suites from here to increase the coverage over time.
<!-- Reviewable comment -ObnHfFn9kK3erCo5Y76:bxqk1s9 -->
<!-- Sent from Reviewable.io -->

spilchen commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1658,6 +1771,28 @@ func (r *testRunner) postTestAssertions(
+				defer db.Close()
+
+				// Use 80% of timeout budget for INSPECT
+				inspectTimeout := time.Duration(float64(timeout) * 0.8)
```
This is just a time budget. The actual INSPECT time could take much less time. I think I would want to see the amount of time it takes to do INSPECT as a validation step over different tests before deciding to change the overall timeout.
<!-- Reviewable comment -ObnIFN95HKlZqZN_vyO:b7ohas0 -->
<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155438)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
i just had one nit that i think we should resolve, but lgtm after that!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-:-Obnk0cd0ZXywJaX-S0A:b-n33y2q)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @herkolategan)

___
*[`pkg/cmd/roachtest/roachtestutil/validation_check.go` line 262 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155438#-ObnhvarBg3wijU-bTBP:-ObnhvarBg3wijU-bTBQ:b84wt81) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f7621e2d1b59d74fd96ad81aa79a5d648ebf67/pkg/cmd/roachtest/roachtestutil/validation_check.go#L262)):*
> ```Go
> 			var stmtErr error
> 			for _, stmt := range statements {
> 				if _, err := db.ExecContext(ctx, stmt); err != nil {
> ```

super nit: to be extra sure that the session variables are applied correctly, we could use `db.Conn` to get a connection handle here. `db` is technically a connection pool, so there's a small chance different connections could be used when sending the different statements in the loop, and in that case the wrong session variables would be set.

 we'd need to make sure to add a deferred `conn.Close()` call to return the connection to the pool when the goroutine exits. `RESET statement_timeout` could also be in a defer, which might make the error handling simpler since we could avoid the need for `stmtErr`


<!-- Sent from Reviewable.io -->

spilchen committed:
roachtest: add INSPECT DATABASE post-validation check

Add INSPECT DATABASE as a post-validation step in selected roachtests to
verify consistency.

Initially enabled only for alterpk* and schemachange* tests, owned by
SQL Foundations. Tests that already run INSPECT internally explicitly
skip this step. A regex-based blacklist disables it for most existing
tests to allow a gradual rollout and avoid destabilizing test runs.

Controlled via a new PostValidation flag. Future tests are expected to
enable this by default.

Fixes #154049
Epic: CRDB-30356
Release note: none
